sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala

```diff
@@ -66,12 +68,16 @@ case class CoalesceShufflePartitions(conf: SQLConf) extends Rule[SparkPlan] {
       val distinctNumPreShufflePartitions =
         validMetrics.map(stats => stats.bytesByPartitionId.length).distinct
       if (validMetrics.nonEmpty && distinctNumPreShufflePartitions.length == 1) {
+        // To avoid perf regression, we by default set the minimum number of coalesced partitions
+        // to Spark default parallelism, so that we don't keep CPU cores idle.
+        val minPartitionNum = conf.getConf(SQLConf.COALESCE_PARTITIONS_MIN_PARTITION_NUM)
+          .getOrElse(session.sparkContext.defaultParallelism)
         val partitionSpecs = ShufflePartitionsUtil.coalescePartitions(
           validMetrics.toArray,
           firstPartitionIndex = 0,
           lastPartitionIndex = distinctNumPreShufflePartitions.head,
           advisoryTargetSize = conf.getConf(SQLConf.ADVISORY_PARTITION_SIZE_IN_BYTES),
-          minNumPartitions = conf.minShufflePartitionNum)
+          minNumPartitions = minPartitionNum)
```

Comment:
So, according to the PR description, do we still need to coalesce if the number of partitions has already been less than `minPartitionNum`?

Reply:
We don't. We should make it clear in `ShufflePartitionsUtil`

Reply:
After looking into `ShufflePartitionsUtil`, we may still do coalesce if the partition sizes are really really small. The target size is at least 16 bytes, so if all partitions are 0 size, we still coalesce.

I don't know why we pick 16, but this makes sense to be as it's obvious an overhead if we launch a task to read just 16 bytes.

Reply:
I see.